todo-mvc-kitchensink/src/widgets/App.ts

```diff
@@ -0,0 +1,194 @@
+	}
+
+	render() {
+		const {
+			pirateTheme = false,
+			theme,
+			activeView = 'list',
+			activeFilter = 'all',
+			showDetails = ''
+		} = this.properties;
+
+		const widgets = [
+			w(ThemeSwitcher, {
+				theme: this.properties.theme,
+				wantsPirate: pirateTheme,
+				onChange: this.changeTheme
+			}),
+			w(Home, {
+				theme,
+				updated: this._updated,
+				todos: this._todos,
+				todo: this._todoItem,
+				completedCount: this._completedCount,
+				search: this._search,
+				activeView,
+				activeFilter,
+				removeTodo: this._removeTodo,
+				toggleTodo: this._toggleTodo,
+				toggleAllTodos: this._toggleAllTodos,
+				clearCompleted: this._clearCompleted,
+				updateTodoItem: this._updateTodo,
+				addTodo: this._setTodo,
+				updateSearch: this._updateSearch,
+				showTodoDetails: this._showTodoDetails
+			})
+		];
+
+		if (showDetails && this._todos.get(showDetails)) {
+			widgets.push(w(TodoDetails, {
+				theme,
+				todo: this._todos.get(showDetails)!,
+				updateTodo: this._setTodo,
+				showTodoDetails: this._showTodoDetails,
+				activeView,
+				activeFilter
+			}));
+		}
+
+		return v('div', [
+			v('section', {
+				classes: this.classes(styles.todoapp)
+			}, widgets),
+			w(AppFooter, {
+				theme: this.properties.theme
+			})
+		]);
+	}
+
+	private _removeTodo(id: string) {
+		const todo = this._todos.get(id);
+
+		if (todo) {
+			this._todos.delete(id);
+			todo.completed && --this._completedCount;
+			this._onUpdate();
+		}
+	}
+
+	private _toggleTodo(id: string) {
+		const todo = this._todos.get(id);
+
+		if (todo) {
+			const completed = !todo.completed;
+			completed ? ++this._completedCount : --this._completedCount;
+			this._setTodo({ completed }, id);
+		}
+	}
+
+	private _toggleAllTodos() {
+		const completed = this._completedCount !== this._todos.size;
+		this._todos.forEach((todo, key) => {
+			this._setTodo({ completed }, key);
+		});
+		this._completedCount = completed ? this._todos.size : 0;
+	}
+
+	private _clearCompleted() {
+		this._todos.forEach((todo, key) => {
+			if (todo.completed) {
+				this._todos.delete(key);
+			}
+		});
+		this._completedCount = 0;
+		this._onUpdate();
+	}
+
+	private _updateTodo(todo: string) {
+		this._todoItem = todo;
+		this.invalidate();
+	}
+
+	private _setTodo(todo: Partial<Todo>, id?: string) {
+		if (!id) {
+			id = uuid();
+			this._todoItem = '';
+		}
+
+		if (todo.label) {
+			todo.label = todo.label.trim();
+			if (!todo.label) {
+				this._removeTodo(id);
+				return;
+			}
+		}
+
+		this._todos.set(id, assign(<any> { id }, this._todos.get(id) || {}, todo));
+		this._onUpdate();
+	}
+
+	private _updateSearch(search: string) {
+		this._search = search;
+		this._onUpdate();
+	}
+
+	private _onUpdate() {
+		this._updated = uuid();
+		this.invalidate();
+	}
+
+	private _showTodoDetails(id: string) {
+		if (id) {
+			document.location.href = `#/${this.properties.activeFilter || 'all'}/todos/${id}?view=${this.properties.activeView || 'list'}`;
```

Comment:
Yeah we need to do something with the router for sure. I wasn't sure what to do here because Ant mentioned the routes basically access the app's public API, which, to me, implies that the app shouldn't have to know about the router either.  Not that I have a good solution for it..